harness/tests/experiment/pytorch/test_reducer.py

```diff
@@ -12,3 +23,96 @@ def test_reducer() -> None:
 
     batches_per_process = [1, 2, 5, 4, 5, 6]
     assert np.around(_reduce_metrics(Reducer.AVG, metrics, batches_per_process), decimals=2) == 6.43
+
+
+DummyTrial = namedtuple("DummyTrial", "distributed_context reducer_context wrapped_reducer")
+
+
+def dummy_reducer(values: List) -> Any:
+    logger.info(f"reducing {values}")
+    flat = [v for sublist in values for v in sublist]
+    return {"values": flat, "sum": sum(flat)}
+
+
+@pytest.mark.parametrize("cross_size", [1, 3])
+@pytest.mark.parametrize("local_size", [1, 3])
+def test_custom_reducer_slot_order(cross_size: int, local_size: int) -> None:
+    size = cross_size * local_size
+    dataset_size = 47
+
+    def do_parallel(fn: Callable) -> List:
+        """
+        Run the same function on one-thread-per-rank, assert there were no exceptions, and return
+        the results from each rank.
+        """
+        results = [None] * size  # type: List
+        errors = [None] * size  # type: List
+        threads = []
+
+        for cross_rank, local_rank in itertools.product(range(cross_size), range(local_size)):
+            rank = cross_rank * local_size + local_rank
+
+            def _fn(rank: int, cross_rank: int, local_rank: int) -> None:
+                try:
+                    results[rank] = fn(rank, cross_rank, local_rank)
+                except Exception:
+                    errors[rank] = traceback.format_exc()
+                    raise
+
+            threads.append(threading.Thread(target=_fn, args=(rank, cross_rank, local_rank)))
+
+        for thread in reversed(threads):
```

Comment:
why reversed()?

Reply:
I want the test to catch a possible incorrect implementation of `_zmq_allgather`.
 
If all operations are started in order, even on separate threads, there is a higher chance that the results of `gather_with_polling` will arrive in order as well. (The order is random but often correct). Then it may _appear_ that we can do without sorting the content reported by workers in `zmq_allgather`. By starting threads in reverse I am making it very unlikely that "worker stuff" will be gathered in order, thereby making it much more likely that the test will fail of the implementation is incorrect.
In other words, if someone screws up the implementation of `_zmq_allgather` by removing the sorting, the test will start failing intermittently, with or without `reverse`, but with `reverse` it will be failing most of the time.

I understand this is a mouthful, and maybe this `reversed` introduces more confusion than benefit.

Reply:
Oh, of course. No this makes perfect sense.

Maybe a comment that says "encourage the allgather to occur in not-the-correct order to test the reordering" or something.